[v14] build: Use buildx and caching for buildboxes, push to ghcr.io

camscale opened the pull request:
Explicitly use `docker buildx build` (instead of `DOCKER_BUILDKIT=1`) to
build the buildbox images, use `--cache-to type=inline` for caching
instead of pulling the image first on CI, and push the image to the
registry if `PUSH=1`.

Push the buildbox images to ghcr.io as our single source of buildbox
images. branch/v15 and master use ghcr.io.

This is a roll-up of changes made iteratively on master / branch/v15
where we have arrived at a solution that seems to build all buildboxes
with any buildx builder and to cache effectively so we reduce build
times. This aligns the branch/v14 build of the buildboxes with the later
branches.

Backport: https://github.com/gravitational/teleport/pull/34950 (partial)
Backport: https://github.com/gravitational/teleport/pull/37559 (partial)
Issue: https://github.com/gravitational/teleport/issues/34281
Test-run: https://github.com/gravitational/teleport.e/actions/runs/8165487223 (build)
Test-run: https://github.com/gravitational/teleport.e/actions/runs/8167190984 (cached)

camscale commented:
There's still a few things on this PR to be fixed so I've put it in draft for now. There are some more parts of #34950 to be backported.

wadells reviewed (commented):
Leaving a comment to take this off my review queue.  Please re-request my review when it is ready for a look.

camscale committed:
build: Cache buildbox layers inline

Use the `docker buildx` `--cache-to type=inline` to put the build cache
in the buildbox image, to with with the existing `--cache-from`
parameter. To do this we need to use `docker buildx`, and to ensure that
a buildbox image is available locally if a container builder is being
used, ensure `--load` is used if we are not pushing to a remote
registry, based on the `PUSH` make variable.

This rolls up a number of changes made on master that are intertwined
with other changes we do not want to backport. This smaller set of
changes aligns the way the buildboxes are built, cached and pushed
across the branches without bringing across the larger scale changes for
PIC/PIE builds and ARM64 FIPS builds.

This also allows us to remove the `PULL_ON_CI` make "macro" since docker
will effectively do the same with `--cache-from` when those images have
an inline cache.

camscale commented:
I've revved this to actually remove some stuff rather than adding more. I was going to try to align v14 with master WRT the dockerfiles, but there are too many changes and too much validation required for now that instead I have backed out the alignment of buildbox naming WRT architecture and have stuck with what v14 currently has.

As a result, all this PR does is ensure the caching is performed usefully using `docker buildx` and `--cache-to type=inline` and to use ghcr.io instead of AWS ECR.

I've tested that the build does not break for release builds and have also re-run the pending build-buildboxes workflow for v14 and it all works as expected.

So, back out of draft and pushing on. v13 is much the same as v14 (with some minor differences with buildbox-arm) so I'll be backporting this to v13.

camscale commented:
Linux release build tests:
* amd64: https://github.com/gravitational/teleport.e/actions/runs/8549686639
* arm64: https://github.com/gravitational/teleport.e/actions/runs/8549827413
* amd64-fips: https://github.com/gravitational/teleport.e/actions/runs/8549961389
* arm: https://github.com/gravitational/teleport.e/actions/runs/8549959945
* 386: https://github.com/gravitational/teleport.e/actions/runs/8549960338

build-buildboxes test run for v14: 
* https://github.com/gravitational/teleport.e/actions/runs/8538231210

camscale commented:
@wadells This is ready for review again.
